박스오피스 [Step4] Mary, idinaloq

idinaloq opened the pull request:
@yim2627
지성 안녕하세요~
STEP4 PR 합니다. 
이번 프로젝트 보기에는 쉬워보였는데, 이전 STEP3 수정과 함께 STEP4가 작성되었고 생각보다 어려워서 오래 걸렸습니다 😩

이번 PR도 잘 부탁드립니다🫡


## 🔍 고민되었던 점
### 1. LoadingView
- 데이터를 다운로드할 때 화면에 indicator만 표시해 주었던 이전 방법과는 다르게 화면 전체를 덮는 LoadingView를 새로 구현해주었습니다.
    ``` swift
    final class LoadingView: UIView
    ```
- 해당 방법으로 이전 스텝에서 발생하였던 문제(다운로드 완료 전까지 셀의 구분선이 화면에 먼저 표시되는 현상)도 해결하였으며, 새로운 ViewController에서도 활용할 수 있었습니다.

### 2. ScrollView ambiguous
- 처음에 scrollView의 autoLayout을 다음과 같이 설정하였습니다. 
    - stackView의 top은 imageView의 bottom으로 잡아주었습니다.
    - stackView의 bottom은 scrollView의 크기에따라 달라질 것이기 때문에 contentLayoutGuide에 잡아주었습니다.
    - scrollView의 가로 스크롤을 막기 위해 화면의 가로 사이즈인 FrameLayoutGuide에 leading과 trailing을 잡아주었습니다.
``` swift
NSLayoutConstraint.activate([
    ...
    contentStackView.topAnchor.constraint(equalTo: imageView.bottomAnchor),
    contentStackView.leadingAnchor.constraint(equalTo: frameLayoutGuide.leadingAnchor),
    contentStackView.trailingAnchor.constraint(equalTo: frameLayoutGuide.trailingAnchor),
    contentStackView.bottomAnchor.constraint(equalTo: contentLayoutGuide.bottomAnchor)
])
```
- 하지만 View Hierarchy에 다음과 같은 경고가 발생하였습니다.
`Scrollable content size is ambiguous for MovieInformationScrollView.`

- 아래 코드와 같이 widthAnchor를 scrollView의 frameLayoutGuide에 잡아주었더니 경고는 해결되었습니다.
이 부분에 대한 공부가 더 필요할 것 같습니다. 😭📚
``` swift
NSLayoutConstraint.activate([
    ...
    contentStackView.topAnchor.constraint(equalTo: imageView.bottomAnchor),
    contentStackView.leadingAnchor.constraint(equalTo: contentLayoutGuide.leadingAnchor),
    contentStackView.trailingAnchor.constraint(equalTo: contentLayoutGuide.trailingAnchor),
    contentStackView.bottomAnchor.constraint(equalTo: contentLayoutGuide.bottomAnchor),
    contentStackView.widthAnchor.constraint(equalTo: frameLayoutGuide.widthAnchor)
])
```

## 🙏 조언을 얻고 싶은 부분
### 1. 여러개의 비동기 작업 끝나는 시점
- `MovieInformationViewController`클래스에서 `receiveImageData()`메서드와 `receiveBoxOfficeData()` 메서드에서 비동기로 데이터를 처리하고 있습니다. 
뷰가 업데이트가 되는 시점(로딩뷰가 사라지는 시점)이 두 비동기 작업 처리가 끝날 때 이루어져야 하기 때문에 방법 중 하나로는 `count`를 지정해서 프로퍼티 옵저버로 로딩 뷰가 사라지도록 설정했습니다.
`dispatch Queue`의 group과 다른 방법들을 찾아보았는데 명확한 해답을 찾을 수 없었습니다.
여러개의 비동기 처리 작업이 끝난 것을 체크하는 방법이 따로 존재하는지 궁금합니다.

### 2. 프로젝트에서의 코드 재사용성
- 프로젝트 내용 구현에만 너무 힘을 쏟은 나머지 `MovieInformationViewController`클래스의 메서드 재사용성에 대한 문제가 있다고 생각합니다. 물론 수정을 하면 좋다고 생각합니다. 하지만 구현 하는것이 먼저라고 생각했고 수정하기에는 시간이 많이 필요할 것 같습니다. 때문에 이렇게 PR을 하게 되었습니다.

앞으로 남은 STEP들도 있을텐데, 이런식으로 하는 방법이 괜찮은 방법일까요...? 죄송합니다 😭 😭 😭 😭

### 3. 셀 재사용 시 구분선이 사라지는 현상
- 현재 첫 번째 박스오피스 리스트인 `DailyBoxOfficeViewController`에서 위아래로 스크롤하여 셀이 사라지게 하면 구분선이 간혹 사라지는 현상이 있습니다.
셀의 윗 부분에만 구분선을 그려지게 한 상태이고, 셀을 새로고침 하게 되면 다시 생기는데 셀의 재사용성과 관련이 있는걸까요...? 
셀이 재사용과 관련이 있다면 셀의 윗부분에만 계속 재사용되어야 하는게 아닌가 라는 생각이 들어 사라지지 않는게 맞다고 생각하는데 사라지고 또 새로고침을 하면 괜찮아 져서 어떤 이유때문에 이런 현상이 발생하는지 모르겠습니다.. 
- 참고로, 이전 스텝에서 피드백주신 대로 영화 제목 길이에 따라 cell의 높이를 유동적으로 변경해주었습니다. 이 부분도 해당 문제와 관련이 있을지 모르겠습니다 😭

<img src="https://hackmd.io/_uploads/SkOiV1-hh.png" width="300">

MaryJo-github committed:
:sparkles: StoryBoard 삭제 및 CollectionView 생성

idinaloq committed:
✨feat: 커스텀 셀 DailyBoxOfficeCollectionViewCell 타입 추가

MaryJo-github committed:
:sparkles: Cell 추가 및 autolayout 설정

idinaloq committed:
✨fear: Date+ 추가

이전 날짜로 박스오피스 받아오도록 수정
박스오피스 정보 받아오도록 수정

idinaloq committed:
📝docs:README 수정

idinaloq committed:
Update README.md

MaryJo-github committed:
:sparkles: Feat: refreshControl 추가

viewController 코드 리팩토링

idinaloq committed:
✨feat: 첫 로딩 화면 구현

activityIndicator

MaryJo-github committed:
:sparkles: Feat: Accessory 추가 및 Label 색상 변경

MaryJo-github committed:
:recycle: Refactor: autolayout 오류 수정 및 셀 초기화 추가

DateManager 타입 생성 및 코드 리팩토링

idinaloq committed:
♻️refactor: 에러처리, 폴더분할, 접근제어자, 오타 수정

final키워드 추가, identifier extension으로 수정

MaryJo-github committed:
:sparkles: Feat: array subscript 구현 및 동적 cell 높이 구현

idinaloq committed:
✨feat: MovieInformation스크롤뷰 추가, MovieInformation뷰컨트롤러 추가

navigationController.pushViewController로 데이터 넘겨주는부분 추가

MaryJo-github committed:
:sparkles: Feat: movieInformation 데이터 다운로드 및 처리부 구현

idinaloq committed:
✨feat: 이미지 다운로드 및 뷰에 표시하도록 추가

loadingView 구현
URL 관련 타입들 URLRequest로 변경
KakaoAPI 이미지 DTO 추가
refresh 로직 수정

MaryJo-github committed:
:sparkles: Feat: LoadingView 추가 및 image 다운로드 실패시 오류처리 구현

MaryJo-github committed:
:adhesive_bandage: Chore: 컨벤션 수정

yim2627 commented on BoxOffice/Controller/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    @objc func refreshData() {
+        receiveData()
+    }
+}
```
```suggestion
    @objc private func refreshData() {
        receiveData()
    }
}
```

yim2627 commented on BoxOffice/Controller/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        }
+        
+        guard let data = boxOfficeData,
+              let data = data.boxOfficeResult.dailyBoxOfficeList[index: indexPath.item] else {
```
👍👍

yim2627 commented on BoxOffice/Controller/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        let cell = DailyBoxOfficeCollectionViewCell(frame: CGRect(x: 0, y: 0, width: width, height: height))
+        
+        cell.titleLabel.text = data.movieName
+        cell.setNeedsLayout()
+        cell.layoutIfNeeded()
```
이 두 메서드는 어떤 역할을 할까요?

idinaloq replied:
[setNeedsLayout](https://developer.apple.com/documentation/uikit/uiview/1622601-setneedslayout)
- setNeedsLayout() 메소드를 호출하면 해당 객체의 레이아웃이 현재의 레이아웃에서 변경되었음을 시스템에 알리며, 이에 대한 업데이트가 다음 업데이트 주기에서 수행될 수 있도록 트리거하는 역할을 합니다.

[layoutIfNeeded](https://developer.apple.com/documentation/uikit/uiview/1622507-layoutifneeded)
- layoutIfNeeded() 메소드를 호출하면 즉각적으로 업데이트를 시작하는 역할을 합니다.

yim2627 replied:
둘다 업데이트를 하는 것이지만, 업데이트 시점에 대한 차이가 있는데 이걸 둘 다 호출 해주신 이유가 있을까요?

MaryJo-github replied:
저희가 제대로 알지 못했던 것 같습니다.
`layoutIfNeeded()`만 호출하여도 레이아웃 업데이트를 즉시 시작하므로 `setNeedsLayout()`을 호출할 필요는 없습니다. 해당 내용 수정하였습니다!

yim2627 commented on BoxOffice/Controller/MovieInformationViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    private var imageSearch: ImageSearch?
+    private let loadingView: LoadingView = LoadingView()
+    private var completionCount: Int = 0 {
+        didSet {
+            if completionCount == 2 {
+                loadingView.hide()
+            }
+        }
```
오 옵저버까지! 👍

yim2627 commented on BoxOffice/Controller/MovieInformationViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    private func receiveBoxOfficeData() {
+        guard let urlRequest = receiveBoxOfficeURLRequest() else { return }
+        
+        NetworkService().fetchData(urlRequest: urlRequest) { result in
```
인스턴스 프로퍼티로 있는데 또 생성해주신 이유가 있을까요?

idinaloq replied:
인스턴스 프로퍼티를 생성하고, 실수로 사용하지 않았던 것으로 확인 되었습니다 😭

yim2627 commented on BoxOffice/Controller/MovieInformationViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    }
+    
+    private func downloadImage() throws -> UIImage {
+        guard let imageURL = imageSearch?.documents[0].imageURL,
```
documents 배열에서도 subscript를 사용할 수 있을 것 같습니다.

idinaloq replied:
```swift
...
guard let imageURL = imageSearch?.documents.first?.imageURL,
...
```
해당 부분 위와 같이 first로 접근하도록 변경했습니다.

yim2627 commented on BoxOffice/Extension/Array+.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by Idinaloq, MARY on 2023/08/08.
+//
+
+extension Array {
+    subscript(index index: Int) -> Element? {
+        return indices ~= index ? self[index] : nil
+    }
+}
```
👍👍

yim2627 commented on BoxOffice/Extension/CALayer+.swift:
```diff
@@ -0,0 +1,19 @@
+import UIKit
+
+extension CALayer {
+    func addSeparator() {
+        let separator = CALayer()
+        
+        separator.frame = CGRect.init(x: 0, y: 0, width: frame.width, height: 0.5)
+        separator.backgroundColor = UIColor.systemGray2.cgColor
+        
+        self.addSublayer(separator)
+    }
```
너무 제한적으로 사용될 것 같습니다. 재사용성을 높일 수 있을 것 같아요.

idinaloq replied:
```swift
  func addSeparator(x: CGFloat, y: CGFloat, width: CGFloat, height: CGFloat) { ... }
```
매개변수로 `x`, `y`, `width`, `height`을 받도록 변경했습니다.

yim2627 commented on BoxOffice/Model/DateManager.swift:
```diff
@@ -0,0 +1,27 @@
+            return formattedYesterday
+        }
+        
+        return "00000000"
```
이전에 오류처리로 변경해주신다 했는데 바뀌질 않았네요

idinaloq replied:
`DateError`타입을 만들어서 에러 처리하도록 변경하였습니다.
해당 부분 수정한다고 해놓고 수정을 안했네요...

yim2627 commented on BoxOffice/Network/NetworkService.swift:
```diff
@@ -16,8 +16,8 @@ final class NetworkService {
 
-    func fetchData(url: URL, completion: @escaping NetworkResult) {
-        let task: URLSessionDataTask  = session.dataTask(with: url) { data, response, error in
+    func fetchData(urlRequest: URLRequest, completion: @escaping NetworkResult) {
+        let task: URLSessionDataTask  = session.dataTask(with: urlRequest) { data, response, error in
```
URLRequest로 변경하신 이유가 있을까요?

idinaloq replied:
kobis에서 데이터를 요청할 때는 `URL`만 있어도 됐지만, kakao에서 받아올 때는 `HTTP`메서드 `"GET"`과 헤더도 추가를 해야하기 때문에 `NetworService`타입을 공통으로 쓰기 위해서 `URLRequest`를 사용하도록 변경했습니다.

yim2627 commented on BoxOffice/View/DailyBoxOfficeCollectionViewCell.swift:
```diff
@@ -0,0 +1,167 @@
+import UIKit
+
+final class DailyBoxOfficeCollectionViewCell: UICollectionViewListCell {
+    let titleLabel: UILabel = {
+        let label: UILabel = UILabel()
+        label.font = .preferredFont(forTextStyle: .title2)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.numberOfLines = 0
+        
+        return label
+    }()
```
얘만 접근제한자가 없는 이유가 있을까요?

idinaloq replied:
`DailyBoxOfficeViewController`의 `collectionView(...) -> CGSize` 메서드에서 현재 셀의 크기를 동적으로 만들어 주고 있습니다.
이 메서드에서 크기를 구하기 위한 셀을 만들어주고, 여기에 `titleLabel`에 들어가는 `movieName`을 할당해 준 다음 `intrinsicContentSize`로 컨텐츠의 크기를 계산해서 셀의 크기를 최종적으로 만들고 있습니다.

때문에 `titleLabel`에 접근제어자가 없는 상태입니다.

yim2627 commented on BoxOffice/View/LoadingView.swift:
```diff
@@ -0,0 +1,53 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    func hide() {
+        DispatchQueue.main.async { [weak self] in
+            self?.layer.zPosition = 0
+            self?.activityIndicator.stopAnimating()
+        }
+    }
```
👍👍

yim2627 reviewed (commented):
안녕하세요 Idinaloq, MARY
이번 스텝도 고생 많으셨습니다. 전체적으로 깔끔하게 잘 해주신 것 같아요.
질문과 코드는 아래에서 다루겠습니다.

1. 여러개의 비동기 작업 끝나는 시점

하나의 그룹으로 묶어서 그룹내의 작업이 모두 끝났다면 notify 등을 통해 작업하는 방법도 있을 것 같습니다. 지금 해주신 프로퍼티 옵저버를 통해서 구현해주신 것도 나쁘진 않은 것 같아요. 다만 ThreadSafe한지 아닌지를 고민해보시고 구현하신다면 더 좋을 것 같습니다.

2. 프로젝트에서의 코드 재사용성

지금 테스트코드 작성도 밀려있다보니, 걱정이 많으실 것 같은데, 야곰도 항상 강조하시듯 진도 나가는게 중요한게 아닌 정확히 알고 넘어가는 것이 중요하다는 것만 아시고 진행하시면 될 것 같습니다. 웬만하면 이번 스텝에서 메서드 재사용성에 대해 생각하신 부분을 적용하시는 게 좋아보여요. 테스트코드도 밀려있으니까요.

3. 셀 재사용 시 구분선이 사라지는 현상

이 부분은.. 저도 정확히는 모르겠지만, 아마 레이아웃에서 나온 문제가 아닐까 생각이 들어요.  레이아웃을 한번 더 셋팅해주는 쪽에서 문제가 생겼을 수 도 있구요.

idinaloq committed:
✨feat: DateError타입 추가

getYesterdayDate()메서드 로직 변경
addSeparator()메서드 매개변수 수정
networkService 인스턴스 프로퍼티 사용하도록 변경

MaryJo-github committed:
:recycle: Refactor: 불필요한 setNeedsLayout 호출부 삭제
